File: web/src/lib/orchestrator.ts
/** Orchestrator — dispatches queries in single or bake-off mode. */

import { arbitrate, type ArbiterResult } from "./arbiter";
import { ENGINES, type Engine, type Length, type QueryRequest, type Speed } from "./config";
import { getQueryFn, type ProviderResponse } from "./providers";

export interface SingleResult {
  kind: "single";
  response: ProviderResponse;
}

export interface BakeoffResult {
  kind: "bakeoff";
  responses: ProviderResponse[];
  arbitration: ArbiterResult;
}

export type QueryResult = SingleResult | BakeoffResult;

export async function runQuery(req: QueryRequest): Promise<QueryResult> {
  if (req.mode === "single") {
    return runSingle(req.prompt, req.engine ?? "claude", req.speed, req.length);
  }
  return runBakeoff(
    req.prompt,
    req.speed,
    req.length,
    req.arbiter ?? "claude"
  );
}

async function runSingle(
  prompt: string,
  engine: Engine,
  speed: Speed,
  length: Length
): Promise<SingleResult> {
  const queryFn = getQueryFn(engine);
  const response = await queryFn(prompt, speed, length);
  return { kind: "single", response };
}

async function runBakeoff(
  prompt: string,
  speed: Speed,
  length: Length,
  arbiterEngine: Engine
): Promise<BakeoffResult> {
  // Fan out to all engines in parallel — tolerate individual failures.
  const settled = await Promise.allSettled(
    ENGINES.map((engine) => getQueryFn(engine)(prompt, speed, length))
  );

  const responses: ProviderResponse[] = [];
  const errors: string[] = [];
  settled.forEach((result, i) => {
    if (result.status === "fulfilled") {
      responses.push(result.value);
    } else {
      errors.push(`${ENGINES[i]}: ${result.reason?.message ?? "unknown error"}`);
    }
  });

  if (responses.length < 2) {
    throw new Error(
      `Too few engines succeeded (${responses.length}/${ENGINES.length}). Failures: ${errors.join("; ")}`
    );
  }

  const arbitration = await arbitrate(prompt, responses, arbiterEngine);

  return { kind: "bakeoff", responses, arbitration };
}

File: metallm/orchestrator.py
"""Orchestrator — dispatches queries in single or bake-off mode."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass

from metallm.arbiter import ArbiterResult, arbitrate
from metallm.config import Engine, Length, Mode, QuerySettings, Speed
from metallm.providers.base import BaseProvider, ProviderResponse
from metallm.providers.chatgpt import ChatGPTProvider
from metallm.providers.claude import ClaudeProvider
from metallm.providers.gemini import GeminiProvider

# ── Provider registry ────────────────────────────────────────────────────

_PROVIDERS: dict[Engine, type[BaseProvider]] = {
    Engine.CLAUDE: ClaudeProvider,
    Engine.GEMINI: GeminiProvider,
    Engine.CHATGPT: ChatGPTProvider,
}


def _get_provider(engine: Engine) -> BaseProvider:
    return _PROVIDERS[engine]()


# ── Result containers ────────────────────────────────────────────────────

@dataclass
class SingleResult:
    response: ProviderResponse


@dataclass
class BakeoffResult:
    responses: list[ProviderResponse]
    arbitration: ArbiterResult


# ── Public API ───────────────────────────────────────────────────────────

async def run(prompt: str, settings: QuerySettings) -> SingleResult | BakeoffResult:
    """Execute a query according to the given settings."""
    if settings.mode == Mode.SINGLE:
        return await _run_single(prompt, settings)
    return await _run_bakeoff(prompt, settings)


async def _run_single(prompt: str, settings: QuerySettings) -> SingleResult:
    assert settings.engine is not None
    provider = _get_provider(settings.engine)
    response = await provider.query(prompt, settings.speed, settings.length)
    return SingleResult(response=response)


async def _run_bakeoff(prompt: str, settings: QuerySettings) -> BakeoffResult:
    assert settings.arbiter is not None

    # Fan out to all three providers in parallel.
    providers = [_get_provider(engine) for engine in Engine]
    tasks = [provider.query(prompt, settings.speed, settings.length) for provider in providers]
    responses: list[ProviderResponse] = await asyncio.gather(*tasks, return_exceptions=False)

    # Arbitrate.
    arbitration = await arbitrate(
        prompt=prompt,
        responses=responses,
        arbiter_engine=settings.arbiter,
    )

    return BakeoffResult(responses=responses, arbitration=arbitration)
